apiserver/facades/agent/uniter/networkinfo.go

```diff
@@ -0,0 +1,458 @@
+			NetworkInfos: []network.NetworkInfo{{Addresses: interfaceAddr}},
+		}
+	}
+
+	for endpoint, space := range bindings {
+		// The binding address information based on link layer devices.
+		info := networkingcommon.MachineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+
+		// Set egress and ingress address information.
+		info.EgressSubnets = endpointEgressSubnets[endpoint]
+
+		ingressAddrs := make([]string, len(endpointIngressAddresses[endpoint]))
+		for i, addr := range endpointIngressAddresses[endpoint] {
+			ingressAddrs[i] = addr.Value
+		}
+		info.IngressAddresses = ingressAddrs
+
+		// If there is no ingress address explicitly defined for a given
+		// binding, set the ingress addresses to either any defaults set above,
+		// or the binding addresses.
+		if len(info.IngressAddresses) == 0 {
+			info.IngressAddresses = defaultIngressAddresses
+		}
+
+		if len(info.IngressAddresses) == 0 {
+			var ingress corenetwork.SpaceAddresses
+			for _, nwInfo := range info.Info {
+				// We need to construct sortable addresses from link-layer
+				// devices, which unlike machine addresses do not have this
+				// information. We can at least ensure that fan addresses will
+				// be sorted after other addresses.
+				scope := corenetwork.ScopeCloudLocal
+				if strings.HasPrefix(nwInfo.InterfaceName, "fan-") {
+					scope = corenetwork.ScopeFanLocal
+				}
+				for _, addr := range nwInfo.Addresses {
+					ingress = append(ingress, corenetwork.NewScopedSpaceAddress(addr.Address, scope))
+				}
+			}
+			corenetwork.SortAddresses(ingress)
+			info.IngressAddresses = make([]string, len(ingress))
+			for i, addr := range ingress {
+				info.IngressAddresses[i] = addr.Value
+			}
+		}
+
+		// If there is no egress subnet explicitly defined for a given binding,
+		// default to the first ingress address. This matches the behaviour when
+		// there's a relation in place.
+		if len(info.EgressSubnets) == 0 && len(info.IngressAddresses) > 0 {
+			var err error
+			info.EgressSubnets, err = network.FormatAsCIDR([]string{info.IngressAddresses[0]})
+			if err != nil {
+				return result, errors.Trace(err)
+			}
+		}
+
+		result.Results[endpoint] = info
+	}
+
+	return result, nil
+}
+
+// getRelationNetworkInfo returns the endpoint name, network space
+// and ingress/egress addresses for the input relation ID.
+func (n *NetworkInfo) getRelationNetworkInfo(
+	relationId int,
+) (string, string, corenetwork.SpaceAddresses, []string, error) {
+	rel, err := n.st.Relation(relationId)
+	if err != nil {
+		return "", "", nil, nil, errors.Trace(err)
+	}
+	endpoint, err := rel.Endpoint(n.unit.ApplicationName())
+	if err != nil {
+		return "", "", nil, nil, errors.Trace(err)
+	}
+
+	pollPublic := n.unit.ShouldBeAssigned()
+	// For k8s services which may have a public
+	// address, we want to poll in case it's not ready yet.
+	if !pollPublic {
+		cfg, err := n.app.ApplicationConfig()
```

Comment:
this logic feels very out of place

Reply:
Yes, for many of these cases we should have separate encapsulations of the CAAS/IAAS functionality for the same end, with a simple choice of which is chosen/injected rather than polluting a single stanza with diverging implementations.

Due to time constraints I have not undertaken that work here.